### PR TITLE
Mthree

### PR DESCRIPTION
**Pull request by rajaiitp**
Implementing the following protocols (check linked issues for descriptions) 

- [Mthree implementation](https://github.com/pasqal-io/qadence-protocols/issues/22)
- [Majority voting](https://github.com/pasqal-io/qadence-protocols/issues/21)
- fixing a bug in tensor_rank_mult to multiply qubit matrices in correct order

### CODE REVIEW

File: qadence_protocols/mitigations/readout.py [hunk 1]
```diff
@@ -7,9 +7,34 @@
...
+) -> npt.NDArray:
+    n_qubits = len(noise_matrices)
+    conf_matrix = csr_matrix((2**n_qubits, 2**n_qubits))
+
+    for j in subspace:
+        for i in subspace:
+            bin_i = bin(i)[2:].zfill(n_qubits)
+            bin_j = bin(j)[2:].zfill(n_qubits)
+
+            ## manually computing the entries of tensor product for only the subspace
```
Comment (RolandMacDoland) [inline qadence_protocols/mitigations/readout.py]:
```suggestion
            # Manually computing the entries of tensor product only for the subspace
```

File: qadence_protocols/mitigations/readout.py [hunk 2]
```diff
@@ -7,9 +7,34 @@
...
+            conf_matrix[i, j] = np.prod(
+                [noise_matrices[k][int(bin_i[k])][int(bin_j[k])] for k in range(n_qubits)]
+            )
+
+        if normalize:
```
Comment (RolandMacDoland) [inline qadence_protocols/mitigations/readout.py]: Are there cases where you _don't_ want to normalise ?

File: qadence_protocols/mitigations/readout.py [hunk 3]
```diff
@@ -98,6 +144,9 @@ def mitigation_minimization(
 
     See Equation (5) in https://arxiv.org/pdf/2001.09980.pdf.
     See Page(3) in https://arxiv.org/pdf/1106.5458.pdf for MLE implementation
+    See Equation (5) in https://arxiv.org/pdf/2108.12518.pdf for MTHREE implementation
+    (matrix free measurement mitigation)
+    This method is supposed to be be reserved for large implementations of 20 plus qubits
```
Comment (RolandMacDoland) [inline qadence_protocols/mitigations/readout.py]:
```suggestion
    Note: This method is supposed to be be reserved for large implementations over 20 qubits.
```

File: qadence_protocols/mitigations/readout.py [hunk 4]
```diff
@@ -116,48 +165,34 @@ def mitigation_minimization(
...
+            p_corr = constrained_inversion(noise_matrices, p_raw)
 
         elif optimization_type == ReadOutOptimization.MLE:
             noise_matrices_inv = list(map(matrix_inv, noise_matrices))
             # Compute corrected inverse using matrix inversion and run MLE.
             p_corr = mle_solve(tensor_rank_mult(noise_matrices_inv, p_raw))
+
+        elif optimization_type == ReadOutOptimization.MTHREE:
+            Confusion_matrix_subspace = subspace_kron(noise_matrices, p_raw.nonzero())
+            # GMRES is best suited for higher dimensional problems
```
Comment (RolandMacDoland) [inline qadence_protocols/mitigations/readout.py]: What is GMRES ?

File: qadence_protocols/mitigations/readout.py [hunk 5]
```diff
@@ -116,48 +165,34 @@ def mitigation_minimization(
...
-            p_corr = res.x
+            p_corr = constrained_inversion(noise_matrices, p_raw)
 
         elif optimization_type == ReadOutOptimization.MLE:
             noise_matrices_inv = list(map(matrix_inv, noise_matrices))
             # Compute corrected inverse using matrix inversion and run MLE.
             p_corr = mle_solve(tensor_rank_mult(noise_matrices_inv, p_raw))
+
+        elif optimization_type == ReadOutOptimization.MTHREE:
+            Confusion_matrix_subspace = subspace_kron(noise_matrices, p_raw.nonzero())
```
Comment (RolandMacDoland) [inline qadence_protocols/mitigations/readout.py]:
```suggestion
            confusion_matrix_subspace = subspace_kron(noise_matrices, p_raw.nonzero())
```

File: qadence_protocols/types.py [hunk 6]
```diff
@@ -25,3 +25,9 @@ class Protocols(StrEnum):
 
 
 qadence_available_protocols = Protocols.list()
+
+
+class ReadOutOptimization(StrEnum):
+    MLE = "mle"
+    CONSTRAINED = "constrained"
+    MTHREE = "mthree"  ## matrix free measurement mitigation
```
Comment (RolandMacDoland) [inline qadence_protocols/types.py]:
```suggestion
    MTHREE = "mthree" 
    # Matrix free measurement mitigation
```

Comment (RolandMacDoland) [inline qadence_protocols/types.py]: Could you also add a comment about the other two ? Thanks.

File: tests/test_readout.py [hunk 7]
```diff
@@ -16,8 +19,17 @@
...
 from qadence_protocols.mitigations.protocols import Mitigations
+from qadence_protocols.mitigations.readout import (
+    matrix_inv,
+    mle_solve,
+    subspace_kron,
+    tensor_rank_mult,
+)
 
 
 @pytest.mark.flaky(max_runs=5)
```
Comment (RolandMacDoland) [inline tests/test_readout.py:38]: I think it does make sense to parametrise over the optimisation types so that all of them are checked and we can also assess the quality of the mitigations methods.

Comment (RolandMacDoland) [reply tests/test_readout.py:38]: Or loop over them in the test itself.

**Review by RolandMacDoland — APPROVED** [review]
Thanks a lot @rajaiitp. Looking good. Few comments from my side.

File: qadence_protocols/mitigations/readout.py (hunk 2 above)
Comment (rajaiitp) [reply qadence_protocols/mitigations/readout.py]: I wanted the function to a be a standalone for just doing subspace kronecker, but didnt want to shift the normalizing outside. Leaving it in this fashion so that anyone who isnt interested in the normalization part can continue to use it. Let me know if this is not a good practice.

File: qadence_protocols/mitigations/readout.py (hunk 4 above)
Comment (rajaiitp) [reply qadence_protocols/mitigations/readout.py]: its an iterative solver for linear equations (Ax = b) suitable for high dimensional problems
changed the comment to "# GMRES (Generalized minimal residual) for linear equations in higher dimension"

File: qadence_protocols/mitigations/readout.py (hunk 2 above)
Comment (RolandMacDoland) [reply qadence_protocols/mitigations/readout.py]: Well, what I understood from my brief overview of the paper is that you always renormalise your subspace. I was just wondering if it made sense to leave the option open.

Comment (rajaiitp) [reply qadence_protocols/mitigations/readout.py]: True, just wanted to keep it for someone who is interested in using kronecker for a subspace. Maybe I will rename the function to reflect it then and remove the optional argument.

File: qadence_protocols/types.py (hunk 6 above)
Comment (rajaiitp) [reply qadence_protocols/types.py]: Moved the comment above and left a line spacing between the variable declarations

File: tests/test_readout.py (hunk 7 above)
Comment (rajaiitp) [reply tests/test_readout.py:38]: Not all the protocols are meant for the same type of problem, but there is some more comparisons that can be done
I will move this to an issue and work on it after the strategies are put in place. 

[Add tests for comparisons of readout methods](https://github.com/pasqal-io/qadence-protocols/issues/20)

**Comment by RolandMacDoland** [comment]
Hey @rajaiitp feels like this can be merged.

**Comment by rajaiitp** [comment]
I wanted to push the changes that you requested in the review along with majority vote, sorry about the delay. I am done, will push in a few mins and request review

File: qadence_protocols/types.py [hunk 8]
```diff
@@ -25,3 +25,17 @@ class Protocols(StrEnum):
 
 
 qadence_available_protocols = Protocols.list()
+
+
+class ReadOutOptimization(StrEnum):
+    # basic inversion and maximum likelihood estimate
+    MLE = "mle"
+
```
Comment (RolandMacDoland) [inline qadence_protocols/types.py]: Remove.

File: qadence_protocols/types.py [hunk 9]
```diff
@@ -25,3 +25,17 @@ class Protocols(StrEnum):
 
 
 qadence_available_protocols = Protocols.list()
+
+
+class ReadOutOptimization(StrEnum):
+    # basic inversion and maximum likelihood estimate
```
Comment (RolandMacDoland) [inline qadence_protocols/types.py]: Capitalise.

File: qadence_protocols/types.py [hunk 10]
```diff
@@ -25,3 +25,17 @@ class Protocols(StrEnum):
...
+    MLE = "mle"
+
+    # constrained inverse optimization
+    CONSTRAINED = "constrained"
+
+    # matrix free measurement mitigation
+    MTHREE = "mthree"
+
+    # majority voting
+    MV = "majority_vote"
```
Comment (RolandMacDoland) [inline qadence_protocols/types.py]:
```suggestion
    MAJ_VOTE = "majority_vote"
```

Or something like this.

File: qadence_protocols/mitigations/readout.py [hunk 11]
```diff
@@ -116,48 +192,42 @@ def mitigation_minimization(
...
             # Compute corrected inverse using matrix inversion and run MLE.
             p_corr = mle_solve(tensor_rank_mult(noise_matrices_inv, p_raw))
+
+        elif optimization_type == ReadOutOptimization.MTHREE:
+            confusion_matrix_subspace = normalized_subspace_kron(noise_matrices, p_raw.nonzero()[0])
+            # GMRES (Generalized minimal residual) for linear equations in higher dimension
+            p_corr, exit_code = gmres(confusion_matrix_subspace, p_raw)
+
+            if exit_code != 0:
+                logger.warning(f"gmres did not converge, exited with code {exit_code}")
```
Comment (RolandMacDoland) [inline qadence_protocols/mitigations/readout.py]:
```suggestion
                logger.warning(f"GMRES did not converge, exited with code {exit_code}")
```

Also: do you want to raise an exception here ?

File: qadence_protocols/mitigations/readout.py [hunk 12]
```diff
@@ -6,10 +6,35 @@
...
 from scipy.optimize import LinearConstraint, minimize
+from scipy.sparse import csr_matrix
+from scipy.sparse.linalg import gmres
+
+from qadence_protocols.types import ReadOutOptimization
+
+logger = get_logger(__name__)
+
+
+def normalized_subspace_kron(noise_matrices: npt.NDArrary, subspace: npt.NDArray) -> npt.NDArray:
```
Comment (RolandMacDoland) [inline qadence_protocols/mitigations/readout.py:21]:
```suggestion
def normalized_subspace(noise_matrices: npt.NDArrary, subspace: npt.NDArray) -> npt.NDArray:
```

Also: could you please add some docstring here ?

File: qadence_protocols/mitigations/readout.py [hunk 13]
```diff
@@ -82,13 +107,58 @@ def renormalize_counts(corrected_counts: npt.NDArray, n_shots: int) -> npt.NDArr
...
+
+    return res.x
+
+
+def majority_vote(noise_matrices: npt.NDArray, p_raw: npt.NDArray) -> npt.NDArray:
+    n_qubits = len(noise_matrices)
+    output = 0
+
+    for i in range(n_qubits):
+        temp = p_raw.reshape([2] * n_qubits)
```
Comment (RolandMacDoland) [inline qadence_protocols/mitigations/readout.py]: I guess `temp` doesn't make for a meaningful name.

File: qadence_protocols/mitigations/readout.py [hunk 14]
```diff
@@ -82,13 +107,58 @@ def renormalize_counts(corrected_counts: npt.NDArray, n_shots: int) -> npt.NDArr
...
+        transpose_axes = [i] + list(range(0, i)) + list(range(i + 1, n_qubits))
```
Comment (RolandMacDoland) [inline qadence_protocols/mitigations/readout.py]:
```suggestion
        transposed_axes = [i] + list(range(0, i)) + list(range(i + 1, n_qubits))
```

File: qadence_protocols/mitigations/readout.py [hunk 15]
```diff
@@ -82,13 +107,58 @@ def renormalize_counts(corrected_counts: npt.NDArray, n_shots: int) -> npt.NDArr
...
+        temp = np.transpose(temp, axes=transpose_axes).reshape(2, 2**n_qubits // 2)
+        probs = np.sum(temp, axis=1)
+
+        # given the output to be 0, the probability of observed outcomes
+        prob_zero = noise_matrices[i][1][0] ** probs[1] * noise_matrices[i][0][0] ** probs[0]
+        # given the output to be 0, the probability of observed outcomes
```
Comment (RolandMacDoland) [inline qadence_protocols/mitigations/readout.py]:
```suggestion
        # Given the output to be 1, the probability of observed outcomes
```

File: qadence_protocols/mitigations/readout.py [hunk 16]
```diff
@@ -82,13 +107,58 @@ def renormalize_counts(corrected_counts: npt.NDArray, n_shots: int) -> npt.NDArr
...
+    normality_constraint = LinearConstraint(np.ones(len(p_raw)).astype(int), lb=1.0, ub=1.0)
+    positivity_constraint = LinearConstraint(np.eye(len(p_raw)).astype(int), lb=0.0, ub=1.0)
+    constraints = [normality_constraint, positivity_constraint]
+    # Minimize the corrected probabilities.
+    res = minimize(corrected_probas, p_corr0, args=(noise_matrices, p_raw), constraints=constraints)
+
+    return res.x
+
+
+def majority_vote(noise_matrices: npt.NDArray, p_raw: npt.NDArray) -> npt.NDArray:
```
Comment (RolandMacDoland) [inline qadence_protocols/mitigations/readout.py:150]: It would have been preferable to do it in another PR for readability.

**Review by RolandMacDoland — COMMENTED** [review]
@rajaiitp Thank you for this. Comments from my side. Question, is the majority voting technique finalised ?

File: qadence_protocols/types.py (hunk 8 above)
Comment (rajaiitp) [reply qadence_protocols/types.py]: I am capitalizing it, not sure exactly what you want me to remove here

Comment (RolandMacDoland) [reply qadence_protocols/types.py]: Ah the blank line. Sorry, been unclear.

File: qadence_protocols/mitigations/readout.py (hunk 11 above)
Comment (rajaiitp) [reply qadence_protocols/mitigations/readout.py]: I think this should be sufficient, I cant think of use case now in this context, can be added if required later

File: qadence_protocols/mitigations/readout.py (hunk 16 above)
Comment (rajaiitp) [reply qadence_protocols/mitigations/readout.py:150]: Ye I started making some changes on it immediately after requesting the first PR. I will ask you about how to handle it in general in our next meeting

File: qadence_protocols/mitigations/readout.py [hunk 17]
```diff
@@ -6,27 +6,61 @@
...
+from scipy.sparse.linalg import gmres
+
+from qadence_protocols.types import ReadOutOptimization
+
+logger = get_logger(__name__)
+
+
+def normalized_subspace_kron(noise_matrices: npt.NDArrary, subspace: npt.NDArray) -> npt.NDArray:
+    """
+    Computes kron within a specified subspace of index locations.
```
Comment (RolandMacDoland) [inline qadence_protocols/mitigations/readout.py]:
```suggestion
    Compute a specified tensor producted subspace of index locations.
```

File: qadence_protocols/mitigations/readout.py [hunk 18]
```diff
@@ -6,27 +6,61 @@
...
+
+    Args:
+        noise: Specifies an array of noise_matrices acting indpedent qubits
+        subspace: List of index locations that defines the subspace for computation
+    """
```
Comment (RolandMacDoland) [inline qadence_protocols/mitigations/readout.py:31]: Return ?

File: qadence_protocols/mitigations/readout.py [hunk 19]
```diff
@@ -6,27 +6,61 @@
...
 def tensor_rank_mult(qubit_ops: npt.NDArray, prob_vect: npt.NDArray) -> npt.NDArray:
-    N = int(np.log2(len(prob_vect)))
     """
     Fast multiplication of single qubit operators on a probability vector.
 
     Similar to how gate operations are implemented on PyqTorch
     Needs to be replaced.
     """
 
+    N = int(np.log2(len(prob_vect)))
```
Comment (RolandMacDoland) [inline qadence_protocols/mitigations/readout.py]: This is the number of qubits right ?

File: qadence_protocols/types.py [hunk 20]
```diff
@@ -25,3 +25,17 @@ class Protocols(StrEnum):
 
 
 qadence_available_protocols = Protocols.list()
+
+
+class ReadOutOptimization(StrEnum):
+    # Basic inversion and maximum likelihood estimate
+    MLE = "mle"
+
```
Comment (RolandMacDoland) [inline qadence_protocols/types.py]: Remove the blank lines. Just to make it more compact.

**Review by RolandMacDoland — APPROVED** [review]
LGTM. I still need to understand a bit more clearly the majority voting technique.

File: qadence_protocols/mitigations/readout.py (hunk 19 above)
Comment (rajaiitp) [reply qadence_protocols/mitigations/readout.py]: renamed to n_qubits

**Comment by rajaiitp** [comment]
Feel free to ping me for a huddle if you want me to walk you through the idea

File: tests/test_readout.py [hunk 21]
```diff
@@ -229,4 +244,151 @@ def test_compare_readout_methods(
...
+        (4, 1),
+        (5, 25),
+        (5, 20),
+        (5, 9),
+        (5, 11),
+        (4, 15),
+        (5, 7),
+    ],
+)
+def test_readout_majority_vote(n_qubits: int, index: int) -> None:
```
Comment (RolandMacDoland) [inline tests/test_readout.py:374]: Last point, if you could implement using the protocol as we discussed, that'd be great. Thanks.

**Review by RolandMacDoland — APPROVED** [review]
LGTM.

File: tests/test_readout.py (hunk 21 above)
Comment (rajaiitp) [reply tests/test_readout.py:374]: I would need to do this for other readout protocols as well I think, let me open an issue and address it separately in the below issue

https://github.com/pasqal-io/qadence-protocols/issues/23
